sphinx/ext/autodoc/preserve_defaults.py

```diff
@@ -165,13 +169,8 @@ def update_defvalue(app: Sphinx, obj: Any, bound_method: bool) -> None:
                             value = ast_unparse(default)
                         parameters[i] = param.replace(default=DefaultValue(value))
 
-            if bound_method and inspect.ismethod(obj):
-                # classmethods
-                cls = inspect.Parameter('cls', inspect.Parameter.POSITIONAL_OR_KEYWORD)
-                parameters.insert(0, cls)
-
             sig = sig.replace(parameters=parameters)
-            if bound_method and inspect.ismethod(obj):
+            if is_classmethod:
                 # classmethods can't be assigned __signature__ attribute.
                 obj.__dict__['__signature__'] = sig
```

Comment:
This is problematic since `obj` might not have `__dict__`. This can happen if it is decorated with a class that uses `__slots__`. It's a pretty cursed example...

```python
from functools import wraps


def make_cursed_decorator(func):
    class CursedDecorator:
        __slots__ = ()

        @staticmethod
        @wraps(func)
        def __call__(*args, **kwargs):
            return func(*args, **kwargs)

    return CursedDecorator()


class A:
    @classmethod
    @make_cursed_decorator
    def foo(cls, x: int, /, *, y: int = 0) -> int:
        return x + y


A.foo(1, y=2)
A.foo.__dict__  # AttributeError
```

Reply:
The outer ``except (AttributeError, TypeError):`` would catch this, but it isn't very clear -- I've added a note to capture this case.